grpc/src/main/java/io/stargate/grpc/service/interceptors/metrics/MetricCollectingServerInterceptor.java

```diff
@@ -0,0 +1,289 @@
+    // or create new impl
+    return Tags.of(
+        TAG_SERVICE_NAME, method.getServiceName(),
+        TAG_METHOD_NAME, method.getBareMethodName(),
+        TAG_METHOD_TYPE, method.getType().name());
+  }
+
+  @Override
+  public <Q, A> ServerCall.Listener<Q> interceptCall(
+      final ServerCall<Q, A> call,
+      final Metadata requestHeaders,
+      final ServerCallHandler<Q, A> next) {
+
+    // TODO: [doug] 2021-05-25, Tue, 11:49 Pass headers or tagProvider
+    final MetricSet metrics = metricsFor(call.getMethodDescriptor());
+    final Consumer<Status.Code> responseStatusTiming =
+        metrics.newProcessingDurationTiming(this.registry);
+
+    final MetricCollectingServerCall<Q, A> monitoringCall =
+        new MetricCollectingServerCall<>(call, metrics.getResponseCounter());
+
+    return new MetricCollectingServerCallListener<>(
+        next.startCall(monitoringCall, requestHeaders),
+        metrics.getRequestCounter(),
+        monitoringCall::getResponseCode,
+        responseStatusTiming);
+  }
+
+  /** Container for all metrics of a certain call. Used instead of 3 maps to improve performance. */
+  protected static class MetricSet {
+
+    private final Counter requestCounter;
+    private final Counter responseCounter;
+    private final Function<Code, Timer> timerFunction;
+
+    /**
+     * Creates a new metric set with the given meter instances.
+     *
+     * @param requestCounter The request counter to use.
+     * @param responseCounter The response counter to use.
+     * @param timerFunction The timer function to use.
+     */
+    public MetricSet(
+        final Counter requestCounter,
+        final Counter responseCounter,
+        final Function<Code, Timer> timerFunction) {
+
+      this.requestCounter = requestCounter;
+      this.responseCounter = responseCounter;
+      this.timerFunction = timerFunction;
+    }
+
+    /**
+     * Gets the Counter that counts the request messages.
+     *
+     * @return The Counter that counts the request messages.
+     */
+    public Counter getRequestCounter() {
+      return this.requestCounter;
+    }
+
+    /**
+     * Gets the Counter that counts the response messages.
+     *
+     * @return The Counter that counts the response messages.
+     */
+    public Counter getResponseCounter() {
+      return this.responseCounter;
+    }
+
+    /**
+     * Uses the given registry to create a {@link Sample Timer.Sample} that will be reported if the
+     * returned consumer is invoked.
+     *
+     * @param registry The registry used to create the sample.
+     * @return The newly created consumer that will report the processing duration since calling
+     *     this method and invoking the returned consumer along with the status code.
+     */
+    public Consumer<Status.Code> newProcessingDurationTiming(final MeterRegistry registry) {
+      final Timer.Sample timerSample = Timer.start(registry);
```

Comment:
Shouldn't this be wrapped in the `Consumer` lambda as well? Right now, the invocation of `newProcessingDurationTiming` will start the timer, but it stops it somewhere in the future when the `Consumer<Status.Code>` is invoked. It may result in wrong incorrect reporting.

Reply:
This is directly from Micrometer, few modifications were made to this class. The plan is to actually drop all this and change what we need directly in upstream. Working on the PR now.